### ISSUE TITLE
When State and Country Picklists are enabled and Country values have been translated, users will receive errors when updating contact addresses

### ISSUE THREAD

**Issue by jamesbsfdc** [SalesforceFoundation/NPSP#6405]
### Background

State and Country picklists let users select States and Countries from predefined, standardized lists, instead of entering State and Country data into text fields. This functionality is not enabled out of the box but can be enabled/disabled by a System Administrator. Administrators can then also translate individual State and Country values using Translation Workbench, for instance for their German users they will want to translate Germany to Deutschland, United States to Vereinigte Staaten von Amerika etc

### Issue Summary

If a customer has State and Country Picklists enabled and has also translated Country values in Translation Workbench they will receive errors or unexpected results when trying to update address information for a contact or in creating a new address record.

The user will also experience unexpected behaviour if they try to update the Billing Address on the Household Account. Typically updating the Billing Address on a Household Account will create a new default address record and update the Mailing Address of the Contact however where the Billing Address is being updated to a translated Country and associated State this updated address will not save to the Contact Mailing Address.

Notes:
- The errors only occur when updating both the State and the Country within the same update i.e a user can successfully update the Country, save the update and then update the State
- The errors do not occur during Contact creation even when setting both the State and Country upon creation
The errors do not occur if when updating Contact Country and State, the Contact Country matches the Country of the Household Account
- The errors do not occur when "Household Account Addresses Disabled" is enabled, nor when updating address information on organisation accounts when "Organizational Account Addresses Enabled" is enabled
- The errors do not occur if the Country is not translated and the error can be replicated in English where English translations have been added (see examples below)

### Steps to Replicate

Enable State and Country Picklists

Ensure English, or the language you are testing as, is active as a Supported Language in Translation Workbench (Setup --> Translation Workbench --> Translation Language Settings)

Add Address Country Translations via Translation Workbench → Translate → Setup Component: Address Country e.g for the language English enter a Address Country Translation of "USA" for the Master Country Label "United States" and "Republic of Ireland" for Ireland

Ensure the Language for the translations you entered in Step 2 matches the Language of the user you are testing as i.e if you entered Country translations in English in Step 2, ensure you are testing as an user whose Language is English

A. Updating a contact who does not have any existing address information

- Create and save a contact without any address information
- Update the same contact created in the previous step, adding both State and Country values for the United States e.g State = Alabama & Country = USA
- User will receive the following error: "Error:A validation rule is preventing the record from saving: A country/territory must be specified before specifying a state value for field"

B. Updating a contact's address by moving them from one translated Country/State to another translated Country/State

- Create and save a contact with a United States address e.g Country = USA & State = Alabama
- Update the contact to have an Irish address e.g Country = Republic of Ireland & State = Carlow
- User will receive the following error: "A validation rule is preventing the record from saving: The BillingStateCode field contains CW, which isn't an available state. Your administrator can add it for you."
- Note: CW is the BillingStateCode for Carlow

C. Try to update a contacts's address by updating the contact's household address

- Create and save a contact with no address
- Navigate to the newly created contact's household account and set a United States address e.g Country = USA & State = Alabama
- The Account update will save successfully, a new address record will be created but the address will not update the Mailing Address of the Contact (a similar update for a Country/State that are not translated will update the contact's address
- Note: If a user now tries to update the Contact address to Country = USA & State = Alabama, the update will save without error, whereas a similar test in A (above) generated an error upon save. Further to this, if after saving the contact update, the user tries to update the contact again to an Irish address e.g Country = Republic of Ireland & State = Carlow, they will again receive the error. If they navigate to the household account again and enter the Country = Republic of Ireland & State = Carlow it will successfully save, and also now allow them to update the contact address to an Irish address

### Potential Workarounds

Disable State and Country Picklists
OR
Disable Household Address in NPSP Add i.e enable "Household Account Addresses Disabled" in NPSP Settings
OR
Remove any Country translations in Translation Workbench
OR
If your users all use the same non-English language, instead of using Translation Workbench to edit Country labels you could edit the Country labels directly in the State and Country picklists area of Setup
OR
Update a Contact Address iteratively e.g where you need to update a Contact's address from Alabama, USA to Carlow, Republic of Ireland, update the country and save the record, then update the State and save the record
OR
Update the Household Account Billing Address first, and then update the Contact's address

**Comment by jamesbsfdc** [comment]
W-9265383

**Comment by davidchengIC** [comment]
@jamesbsfdc  This also occurs when there are no translations, the account's billing address fields are blank, and there is just one Address record and it is seasonal.

This can occur in a data migration where all NPSP triggers were disabled during the migration.

Example: A household account has blank billing address field and a single Address record.  The address is a valid Florida USA address and the Seasonal fields are populated.

When the season address Batch Apex runs, it throws an error:
Update failed. First exception on row 0 with id 0015t000003Cq0nAAC; first error: FIELD_INTEGRITY_EXCEPTION, The BillingStateCode field contains FD, which isn't an available state. Your administrator can add it for you.: [BillingStateCode]

This is because "Florida" is also a department in Uruguay and its ISO code is FD.

**Comment by jamesbsfdc** [comment]
Hi @davidchengIC  Thanks for reporting this. Do you have an open support case about this issue? If so would you mind posting a case number, or if there is no current case would you mind creating one and posting back with the case number? 

If we can verify the issue, we can look at widening the scope of this Github issue or possibly creating a new one. Until address type fields are supported on custom objects I fear there may be no easy way of resolving this and similar issues that does not entail some kind of tradeoff.

**Comment by davidchengIC** [comment]
@jamesbsfdc I don't have a case yet.  I did find a workaround - if I deactivate the Florida value for Uruguay, then the batch will process the Florida USA address successfully.

**Comment by kporray1** [comment]
Please follow this [Known Issue](https://trailblazer.salesforce.com/issues_view?id=a1p4V000002dlBhQAI&title=users-receive-errors-when-updating-contact-addresses-if-state-and-country-picklists-are-enabled-and-country-values-have-been-translated) in the Trailblazer Community for updates.

**Comment by NormCopeland** [comment]
I'm encountering this issue.[ The linked known issue is closed](https://issues.salesforce.com/issue/a028c00000gAwSoAAK/users-receive-errors-when-updating-contact-addresses-if-state-and-country-picklists-are-enabled-and-country-values-have-been-translated), but doesn't seem to indicate that it was fixed. I've tried turning off Translation Workbench since we don't need it, but it did not help. The workarounds (updating address on the account rather than the contact, or updating it by doing the country first, then state/province next (or vice versa)) are not working. Does any one know the status of this issue?
